README.md
[![Build Status](https://travis-ci.org/urbanriskmap/cognicity-server.svg?branch=master)](https://travis-ci.org/urbanriskmap/cognicity-server) [![Coverage Status](https://coveralls.io/repos/github/urbanriskmap/cognicity-server/badge.svg?branch=master)](https://coveralls.io/github/urbanriskmap/cognicity-server?branch=master)

DOI for current [stable release] (https://github.com/urbanriskmap/cognicity-server/releases)


[![DOI](https://zenodo.org/badge/73803254.svg)](https://zenodo.org/badge/latestdoi/73803254)
## cognicity-server
API Server for CogniCity

### Summary

This is the NodeJS server which runs the CogniCity Data API used by Urban Risk Map instances, such as [PetaBencana.id](https://petabencana.id) site.  

### Run
ES6 Support is provided by Babel.

1. Install requirements from the provided `package.json` by doing `npm install`.

2. Copy the `sample.env` file to a local `.env` and fill-in the required parameters. This local file will be ignored by Git and so should be secret safe. Further details on configuration are described below.

3. To run a local development instance of the server do `npm run dev`

### Configuration
Server configuration parameters are stored in a configuration file which is parsed by index.js on startup. Local configuration parameters are imported from the `.env` into `src/config.js`.  See `config.js` for full details example configuration. Any variable not defined in `.env` will pickup the default value below (also see `config.js`)—note that local environment variables will override both `.env` and `config.js`.  The following environment variables are currently supported by the configurtion:

* `APP_NAME`: Name of the application (default: `cognicity-server`)
* `API_FEEDS_QLUE_CITIES`: Names of cities used by the Qlue data feed
* `API_FEEDS_QLUE_DISASTER_TYPES`: Names of disaster types used by the Qlue data feed
* `API_FEEDS_DETIK_DISASTER_TYPES`: Names of disaster types used by the Detik data feed
* `API_REPORTS_TIME_WINDOW`: Time window for report data queries (default 1 hour)
* `API_REPORTS_TIME_WINDOW_MAX`: Maximum limit for time window (default 1 week)
* `API_REPORTS_LIMIT`: Total maximum number of reports to return in a single request
* `API_FLOODGAUGE_REPORTS_TIME_WINDOW`: Time window for flood data (normally 12 hours)
* `API_FLOODGAUGE_REPORTS_TIME_WINDOW`: Total maximum number of flood gauge records to return in a single request
* `AUTH0_AUDIENCE`: Data API to be authenticated
* `AUTH0_CLIENT_ID`: Auth0 client ID (NOTE: this is mandatory and no default value)
* `AUTH0_ISSUER`: Web address of Auth0 instance
* `AWS_REGION`: Region for AWS Infrastructure
* `AWS_S3_ACCESS_KEY_ID`: Access key ID for AWS S3 bucket
* `AWS_S3_SECRET_ACCESS_KEY`: Access key secret for AWS S3 bucket
* `AWS_S3_SIGNATURE_VERSION`: Version of AWS S3 signature to use
* `AUTH0_SECRET`: Auth0 secret (NOTE: this is mandatory and no default value)
* `BODY_LIMIT`: Maximum body size POST/PUT/PATCH (default: `100kb`)
* `CACHE`: Should caching be enabled? (default: `false`)
* `CACHE_DURATION_CARDS`: How long should cards be cached for? (default: '1 minute')
* `CACHE_DURATION_FLOODS`: How long should floods be cached for? (default: '1 hour')
* `CACHE_DURATION_FLOODS_STATES`: How long should flood states be cached for? (default: '1 hour')
* `CACHE_DURATION_INFRASTRUCTURE`: How long should infrastructure be cached for? (default: '1 hour')
* `CAP_DEFAULT_EXPIRE_SECONDS`: Default expire value for CAP output in seconds
* `CAP_TIMEZONE`: Timezone for CAP output
* `COMPRESS`: Should the server gzip compress results? Only works if CACHE is disabled. (default: `false`)
* `CORS`: Should Cross Object Resource Sharing (CORS) be enabled (default: `true`)
* `CORS_HEADERS`: CORS headers to use (default: `[Link]`)
* `DAMAGE_COMPONENT`: Building components for which damage can be reported (default: `roof,walls,plinth,nonstructural`)
* `DISASTER_TYPES`: Disaster type keywords for report classification (default: `flood,prep,assessment`)
* `FORMAT_DEFAULT`: Which format to return results in by default (default: `json`)
* `FORMATS`: Formats supported by the system (as comma separated list) (default: `json,xml`)
* `GEO_FORMAT_DEFAULT`: Which format to return geographic results in by default (default: `topojson`)
* `GEO_FORMATS`: Geographic formats supported by the system (as comma separated list) (default: `topojson,geojson,cap`)
* `GEO_PRECISION`: Precision to use when rounding geographic coordinates (default: `10`)
* `IMAGES_BUCKET`: AWS S3 bucket for image uploads (default: `testing-riskmap-image-uploads`)
* `IMAGES_HOST`: Endpoint for image hosting (default: `images.petabencana.id`),
* `INFRASTRUCTURE_TYPES`: Infrastructure types supported (as comma separated list) (default: `floodgates,pumps,waterways`)
* `LANGUAGES`: Supported languages
* `LOG_CONSOLE`: In development mode we log to the console by default, in other environments this must be enabled if required by setting this parameter to `true` (default: `false`)
* `LOG_DIR`: Which directory should logs be written to.  If blank, not supplied or the directory is not writable by the application this will default to the current directory
* `LOG_JSON`: Should json format be used for logging (default: `false`)
* `LOG_LEVEL`: What level to log at. Levels are: `silly`, `debug`, `verbose`, `info`, `warn`, `error`. `debug` level is recommended for development.  (default: `error`)
* `LOG_MAX_FILE_SIZE`: Maximum size of log file in bytes before rotating (default: `1024 * 1024 * 100` i.e. `100mb`)
* `LOG_MAX_FILES`: Maximum number of log files before rotation (default: `10`)
* `NODE_ENV`: Which environment are we in.  Environments are: development, test, staging, production (default: `development`)
* `PGHOST`: Postgres DB hostname (default: `127.0.0.1`)
* `PGDATABASE`: Postgres DB database name (default: `cognicity`)
* `PGPASSWORD`: Postgres DB password (default: `p@ssw0rd`)
* `PGPORT`: Postgres DB port (default: `5432`)
* `PGSSL`: SSL enabled on Postgres DB connection? (default: `false`)
* `PGTIMEOUT`: Max duration on DB calls before timeout (in milliseconds) (default: `5000` i.e. 5 seconds)
* `PGUSER`: Postgres DB username (default: `postgres`)
* `PORT`: Which port should the application run on (default: `8001`)
* `REGION_CODES`: Which region codes are supported (as comma separated list) (default: `jbd,bdg,sby`)
* `REPORT_TYPES`: Classifiers for report types (default: `drain,desilting,canalrepair,treeclearing,flood,assessment`)
* `RESPONSE_TIME`: Should the server return an `X-Response-Time` header detailing the time taken to process the request.  This is useful for both development to identify latency impact on testing and production for performance / health monitoring (default: `false`)
* `SECURE_AUTH0`: Whether Auth0 JWT token security should be applied to secure routes (default: `false`)
* `TABLE_FLOODGAUGE_REPORTS`: Postgres table name for flood-gauge reports
* `TABLE_FEEDS_QLUE`: Postgres table name for Qlue feed
* `TABLE_FEEDS_DETIK`: Postgres table name for Detik feed
* `TABLE_GRASP_CARDS`: Postgres table name for Grasp Cards
* `TABLE_GRASP_LOG`: Postgres table name for Grasp activity
* `TABLE_GRASP_REPORTS`: Postgres table name for Grasp reports
* `TABLE_INSTANCE_REGIONS`: Postgres table for operating regions
* `TABLE_LOCAL_AREAS`: Postgres table for local areas data for each operating region
* `TABLE_REM_STATUS`: Postgres table for current flood states from REM
* `TABLE_REM_STATUS_LOG`: Postgres table for REM log
* `TABLE_REPORTS`: Postgres table for reports

A few points to note on config:

* AWS Beanstalk: If you're deploying to [AWS Elastic Beanstalk](http://docs.aws.amazon.com/elasticbeanstalk/latest/dg/Welcome.html) you may need to configure environment variables for each of the above under Configuration -> Software configuration.

### Building
Run `npm run -s build` to build.

### Testing
Testing is run by [Travis](https://travis-ci.org/urbanriskmap/cognicity-server). ESLint runs to check syntax. Integration tests, formed by chaining unit tests, are used to check the API.  Coverage is provided by Istanbul and [Coveralls](https://coveralls.io/github/urbanriskmap/cognicity-server). See src/test/ for scripts. Beware that integration tests may pollute tables (e.g. user tables in feeds), it is not recommended to run tests against prod databases with live data. The default database (used for testing) is cognicity. Travis-ci creates a new schema instance for testing using https://github.com/urbanriskmap/cognicity-schema, see the .travis.yml file for more details.

To run tests locally a new database "cognicity_server_testing" is required on localhost.

Code follows the [Google JavaScript style](https://google.github.io/styleguide/jsguide.html).ESLint is run with tests to enfoce style.

### Issue Tracking
Issues are tracked using [GitHub](https://github.com/urbanriskmap/cognicity-server/issues)

### Code documentation

Internal code documentation is in the JSDoc format. Generated documentation is
in the  jsdoc/ folder. To build documentation do:
```sh
npm run jsdoc
```

Docs are hosted at https://urbanriskmap.github.io/cognicity-server/jsdoc/

### Package management
Before deployment:
* Use NVM to switch to node and NPM versions specified in package.json for production
* Run npm install
* Commit changes to package-lock.json

### Release
The release procedure is as follows:
* Update the CHANGELOG.md file with the newly released version, date, and a high-level overview of changes. Commit the change.
* Build code documentation and commit changes.
* Download the latest version of the Swagger API file for include in the release
* Create a tag in git from the current head of master. The tag version should be the same as the version specified in the package.json file - this is the release version.
* Update the version in the package.json file and commit the change.
* Further development is now on the updated version number until the release process begins again.

### API Notes
Full API documentation at https://docs.petabencana.id. This documentation is stored in the [petabencana-docs](https://github.com/urbanriskmap/petabencana-docs) repository.

The swagger files under [/apigw](apigw/) describe the API in [swagger](https://swagger.io/) format (with AWS API Gateway extensions) for each of our deployments. For a new or updated deployment, the references to the (Elastic Beanstalk) hostnames and Lambda Amazon Resource Names will need to be updated first before import, and then permissions to trigger Lambdas granted to the API.

- The dbgeo library expects timestamps from database to be in UTC (i.e. not a local timezone)

### License
See LICENSE.md


CHANGELOG.md
## cognicity-server
API Server for CogniCity

### v3.0.0
* Refactored server just for data API (web content hosted separately)
* Merged REM API functionality into mainline server
* Added authentication to some endpoints using Auth0/JWT
* Added new functionality for data collection using GRASP Cards
* API Documentation moved to dedicated wiki: https://docs.petabencana.id/

### v3.0.1
* Refactored server module for better testability
* Added tests for endpoints using unit.js and mocha
* Added coverage for tests using Istanbull, nyc and coveralls
* Added integration tests on Travis, including against cognicity-schema
* Extended data format of cards data object to collect different types of data
* Merged changes from Chennai deployment, for better deployment in new cities

### v3.0.2
* Updated to use Node v8.1.4 with npm 5.0.3
* Pushed image url assignment to server
* Updated cards/ patch endpoint to stop potential overwriting of image url
* Card IDs generated as UUID by Postgres
* Added schema version to / endpoint
* Image upload changed to use AWS S3 signed links

### v3.0.3
* Added a default expire time to CAP output
* Add /floods/archive endpoint
* Added database log entry for card creation
* Added /floods/timeseries endpoint
* Added /reports/timeseries endpoint
* Added time window check on archive and time series endpoints
* Updated API definitions in swagger file

### v3.0.4
* Added Semarang instance for Indonesian deployment
* Fixed out of date link in README to list of releases

### v3.0.5
* Fixed bug in API (swagger-api.json) for flood archive and timeseries endpoints
* Added PATCH /reports/:id endpoint for "points"
* Updated documentation of Swagger files
* Updated Babel package
* Remove username field from GET /cards/:id
* Fix bug with time series endpoints

### v3.0.6
* Fix bug with expiry time in CAP format of floods/ endpoint
* Updated Moment (known vulnerability in v2.18.1)
* Test against PostgreSQL v10 (PostGIS 2.4)
* Remove ?format option from /reports endpoint
* Change floods endpoint to return new attributes column
* Requires cognicity-schema v3.0.7 or later
* Updated npm packages

### v3.1.0
* Add city query parameter to /floods/timeseries and /floods/archive endpoints
* Add time zone validation to all endpoints accepting timestamps
* Notifications of received reports are now run from the server (added for testing in v3.0.6)
* Add deployments for India

### v3.1.1
* Add /reports/:id/flag endpoint
* Added support for assessment cards

### v3.1.2


src/config.js
/**
 * CogniCity Server configuration
 * @file config
 * @return {Object} Server configuration
*/
/* eslint-disable max-len */
require('dotenv').config({silent: true});

export default {
  APP_NAME: process.env.APP_NAME || 'cognicity-server',
  API_FEEDS_QLUE_CITIES: (process.env.API_FEEDS_QLUE_CITIES || 'jabodetabek,bandung,surabaya').split(','),
  API_FEEDS_QLUE_DISASTER_TYPES: (process.env.API_FEEDS_QLUE_DISASTER_TYPES || 'flood').split(','),
  API_FEEDS_DETIK_DISASTER_TYPES: (process.env.API_FEEDS_DEIK_DISASTER_TYPES || 'flood').split(','),
  API_REPORTS_TIME_WINDOW: process.env.API_REPORTS_TIME_WINDOW || 3600,
  API_REPORTS_TIME_WINDOW_MAX: process.env.API_REPORTS_TIME_WINDOW_MAX || 604800, // 1w
  API_REPORTS_LIMIT: process.env.API_REPORTS_LIMIT,
  API_FLOODGAUGE_REPORTS_TIME_WINDOW: process.env.API_FLOODGAUGE_REPORTS_TIME_WINDOW || 43200,
  API_FLOODGAUGE_REPORTS_LIMIT: process.env.API_FLOODGAUGE_REPORTS_LIMIT,
  AUTH0_AUDIENCE: process.env.AUTH0_AUDIENCE || 'https://data.petabencana.id',
  AUTH0_CLIENT_ID: process.env.AUTH0_CLIENT_ID || 'auth0_client_id',
  AUTH0_ISSUER: process.env.AUTH0_ISSUER || 'https://petabencana.au.auth0.com',
  AUTH0_SECRET: process.env.AUTH0_SECRET || 'secret',
  AWS_REGION: process.env.AWS_REGION || 'ap-southeast-1',
  AWS_S3_ACCESS_KEY_ID: process.env.AWS_S3_ACCESS_KEY_ID || '',
  AWS_S3_SECRET_ACCESS_KEY: process.env.AWS_S3_SECRET_ACCESS_KEY || '',
  AWS_S3_SIGNATURE_VERSION: process.env.AWS_SIGNATURE_VERSION || 'v4',
  BODY_LIMIT: process.env.BODY_LIMIT || '100kb',
  CACHE: process.env.CACHE === 'true' || false,
  CACHE_DURATION_CARDS: process.env.CACHE_DURATION_CARDS || '1 minute',
  CACHE_DURATION_FLOODS: process.env.CACHE_DURATION_FLOODS || '1 hour',
  CACHE_DURATION_FLOODS_STATES: process.env.CACHE_DURATION_FLOODS_STATES || '1 hour',
  CACHE_DURATION_INFRASTRUCTURE: process.env.CACHE_DURATION_INFRASTRUCTURE || '1 hour',
  CAP_DEFAULT_EXPIRE_SECONDS: process.env.CAP_DEFAULT_EXPIRE_SECONDS || 21600,
  CAP_TIMEZONE: process.env.CAP_TIMEZONE || 'Asia/Jakarta',
  COMPRESS: process.env.COMPRESS === 'true' || false,
  CORS: process.env.CORS === 'true' || true,
  CORS_HEADERS: process.env.CORS_HEADERS || ['Link'],
  DAMAGE_COMPONENT: process.env.DAMAGE_COMPONENT || 'roof,walls,plinth,nonstructural'.split(','),
  DISASTER_TYPES: (process.env.DISASTER_TYPES || 'flood,prep,assessment').split(','),
  FORMAT_DEFAULT: process.env.FORMAT_DEFAULT || 'json',
  FORMATS: (process.env.FORMATS || 'json').split(','),
  GEO_FORMAT_DEFAULT: process.env.GEO_FORMAT_DEFAULT || 'topojson',
  GEO_FORMATS: (process.env.GEO_FORMATS || 'geojson,topojson').split(','),
  GEO_PRECISION: process.env.GEO_PRECISION || 10,
  IMAGES_BUCKET: process.env.IMAGES_BUCKET || 'petabencana-image-uploads',
  IMAGES_HOST: process.env.IMAGES_HOST || 'images.petabencana.id',
  IMAGE_MIME_TYPES: (process.env.IMAGE_MIME_TYPES || 'image/png,image/jpeg,image/gif').split(','),
  INFRASTRUCTURE_TYPES: (process.env.INFRASTRUCTURE_TYPES || 'basins,floodgates,pumps,sites,waterways').split(','),
  LANGUAGES: (process.env.LANGUAGES || 'en,id').split(','),
  LOG_CONSOLE: process.env.LOG_CONSOLE === 'true' || false,
  LOG_DIR: process.env.LOG_DIR || '',
  LOG_JSON: process.env.LOG_JSON === 'true' || false,
  LOG_LEVEL: process.env.LOG_LEVEL || 'error',
  LOG_MAX_FILE_SIZE: process.env.LOG_MAX_FILE_SIZE || 1024 * 1024 * 100,
  LOG_MAX_FILES: process.env.LOG_MAX_FILES || 10,
  NODE_ENV: process.env.NODE_ENV || 'development',
  NOTIFY_API_KEY: process.env.NOTIFY_API_KEY || '',
  NOTIFY_ENDPOINT: process.env.NOTIFY_ENDPOINT || 'api.petabencana.id/notify',
  PGHOST: process.env.PGHOST || '127.0.0.1',
  PGDATABASE: process.env.PGDATABASE || 'cognicity',
  PGPASSWORD: process.env.PGPASSWORD || 'p@ssw0rd',
  PGPORT: process.env.PGPORT || 5432,
  PGSSL: process.env.PGSSL === 'true' || false,
  PGTIMEOUT: process.env.PGTIMEOUT || 10000,
  PGUSER: process.env.PGUSER || 'postgres',
  PORT: process.env.PORT || 8001,
  REGION_CODES: (process.env.REGION_CODES || 'jbd,bdg,sby,srg').split(','),
  REPORT_TYPES: (process.env.REPORT_TYPES || 'drain,damage,power,treeclearing,flood,assessment').split(','),
  RESPONSE_TIME: process.env.RESPONSE_TIME === 'true' || false,
  SECURE_AUTH0: process.env.SECURE_AUTH0 === 'true' || false,
  TABLE_FLOODGAUGE_REPORTS: process.env.TABLE_FLOODGAUGE_REPORTS || 'floodgauge.reports',
  TABLE_FEEDS_QLUE: process.env.TABLE_FEEDS_QLUE || 'qlue.reports',
  TABLE_FEEDS_DETIK: process.env.TABLE_FEEDS_DETIK || 'detik.reports',
  TABLE_GRASP_CARDS: process.env.TABLE_GRASP_CARDS || 'grasp.cards',
  TABLE_GRASP_LOG: process.env.TABLE_GRASP_LOG || 'grasp.log',
  TABLE_GRASP_REPORTS: process.env.TABLE_GRASP_REPORTS || 'grasp.reports',
  TABLE_INSTANCE_REGIONS: process.env.TABLE_INSTANCE_REGIONS || 'cognicity.instance_regions',
  TABLE_LOCAL_AREAS: process.env.TABLE_LOCAL_AREAS || 'cognicity.local_areas',
  TABLE_REM_STATUS: process.env.TABLE_REM_STATUS || 'cognicity.rem_status',
  TABLE_REM_STATUS_LOG: process.env.TABLE_REM_STATUS_LOG || 'cognicity.rem_status_log',
  TABLE_REPORTS: process.env.TABLE_REPORTS || 'cognicity.all_reports',
  TABLE_REPORTS_POINTS_LOG: process.env.TABLE_REPORTS_LOG || 'cognicity.reports_points_log',
};
